mne/preprocessing/xdawn.py

```diff
@@ -129,31 +127,143 @@ def least_square_evoked(epochs, return_toeplitz=False):
+    info = cp.deepcopy(info)
     for name, data in evoked_data.items():
-        n_events = len(events[events[:, 2] == epochs.event_id[name]])
-        evoked = EvokedArray(data, info, tmin=epochs.tmin,
+        n_events = len(events[events[:, 2] == event_id[name]])
+        evoked = EvokedArray(data, info, tmin=tmin,
                              comment=name, nave=n_events)
-        evokeds[name] = evoked
+        evokeds[name] = evoked.data
 
     if return_toeplitz:
         return evokeds, toeplitz
 
     return evokeds
 
 
-class Xdawn(TransformerMixin, ContainsMixin):
+class _Xdawn(TransformerMixin, ContainsMixin):
+    """Private class containing common functions for Xdawn and
+       XdawnTransformer.
+    """
+    def __init__(self, n_components, signal_cov, reg):
+        self.n_components = n_components
+        self.signal_cov = signal_cov
+        self.reg = reg
+        self.filters_ = dict()
+        self.patterns_ = dict()
+        self.evokeds_cov_ = dict()
+
+    def _get_signal_cov(self, epochs_data):
+        if self.signal_cov is None:
+            sig_data = np.hstack(epochs_data)
+            self.signal_cov_ = _regularized_covariance(sig_data, self.reg)
+        elif isinstance(self.signal_cov, Covariance):
+            self.signal_cov_ = self.signal_cov.data
+        elif isinstance(self.signal_cov, np.ndarray):
+            self.signal_cov_ = self.signal_cov
+        else:
+            raise ValueError('signal_cov must be None, a covariance instance '
+                             'or a ndarray')
+
+    def _epochs_data_dict(self, epochs_data, y, event_id):
+        """Select index of same event_id from epochs_data and make a dict."""
+        # TODO: Implement pythonic way of writing this.
+        epochs_data_dict = dict()
+        for eid in event_id:
+            index = []
+            for idx in range(len(y)):
+                if y[idx] == event_id[eid]:
+                    index.append(idx)
+            epochs_data_dict[eid] = epochs_data[index]
+
+        return epochs_data_dict
+
+    def _fit_xdawn(self, epochs_data, y, event_id, events=None, tmin=None,
+                   tmax=None, info=None, baseline=None):
+        if self.correct_overlap:
+            if baseline is not None:
+                raise ValueError('Baseline correction must be None if overlap '
+                                 'correction activated')
+            evokeds, toeplitz = least_square_evoked(epochs_data, events,
+                                                    event_id, tmin, tmax,
+                                                    info,
+                                                    return_toeplitz=True)
+        else:
+            epochs_data_dict = self._epochs_data_dict(epochs_data, y,
+                                                      event_id)
+            evokeds = dict()
+            toeplitz = dict()
+            for eid in event_id:
+                evokeds[eid] = np.mean(epochs_data_dict[eid], axis=0)
```

Comment:
The existing API has `dict` although it stores `evoked` data type. 
@kingjr it can be changed to numpy array, although it will involve lot of code change and testing. Does not seem a good idea to me.